### DISCUSSION TITLE
Help with Fixing an Issue

### DISCUSSION THREAD

**Discussion by kelleyma49** [microsoft/terminal#18085] [General]
I have a bug that I'm trying to fix and I'm blocked. Here's the bug: 

https://github.com/kelleyma49/PSFzf/issues/274 

This image from that bug succinctly sums up the issue: 

![image](https://github.com/user-attachments/assets/5d2f143f-4067-451f-9f4c-9a2316ffd2ba)

I believe that a VT command or a Windows command that leaves the Windows terminal in a weird state. I've tried going through all of the VT commands to reset the terminal state, but I can never get the terminal back into the correct state. 

Questions for the Windows terminal engineers: 

- Is there a VT command that might reset all formatting that would resolve the issue?
- Is there a known VT command or Windows function call that would leave the terminal in a state like this?
- Any other advice?

**Comment by o-sdn-o** [comment]
> I believe that a VT command or a Windows command that leaves the Windows terminal in a weird state.

To localize the bug source (someone leaves the terminal with `DISABLE_NEWLINE_AUTO_RETURN` mode enabled), a step-by-step repro from ground (fresh setup) is required.

> - Is there a VT command that might reset all formatting that would resolve the issue?
> - Is there a known VT command or Windows function call that would leave the terminal in a state like this?

You can use Win32 Console API to toggle the `DISABLE_NEWLINE_AUTO_RETURN` mode.
- Powershell script.
  `con_mode.ps1`:
  ```pwsh
  $win32 = Add-Type -Name WinAPI -PassThru -MemberDefinition @'
  [DllImport("kernel32.dll")] public static extern IntPtr GetStdHandle(int nStdHandle);
  [DllImport("kernel32.dll")] public static extern uint SetConsoleMode(IntPtr hConsoleHandle, uint dwMode);
  [DllImport("kernel32.dll")] public static extern uint GetConsoleMode(IntPtr hConsoleHandle, out uint lpMode);
  '@
  $STD_OUTPUT_HANDLE = -11
  $DISABLE_NEWLINE_AUTO_RETURN = 8
  $mode = 0
  $handle = $win32::GetStdHandle($STD_OUTPUT_HANDLE)
  $rc = $win32::GetConsoleMode($handle, [ref]$mode)
  if ($args[0]) { $mode = $mode -bOr $DISABLE_NEWLINE_AUTO_RETURN }
  else          { $mode = $mode -bAnd (-bNot $DISABLE_NEWLINE_AUTO_RETURN) }
  $rc = $win32::SetConsoleMode($handle, $mode)
  ```
- Turn on `DISABLE_NEWLINE_AUTO_RETURN`:
  `pwsh`:
  ```pwsh
  ./con_mode.ps1 1
  "test: line1`nline2`nline3"
  
  ```
- Turn off `DISABLE_NEWLINE_AUTO_RETURN`:
  `pwsh`:
  ```pwsh
  ./con_mode.ps1 0
  "test: line1`nline2`nline3"
  
  ```

Also you can use vt-sequence [LNM (Line Feed/New Line Mode)](https://vt100.net/docs/vt510-rm/LNM.html), xlink: #15167:

- Turn on `DISABLE_NEWLINE_AUTO_RETURN`:
  `pwsh`:
  ```pwsh
  "`e[20l"
  "test: line1`nline2`nline3"
  
  ```
- Turn off `DISABLE_NEWLINE_AUTO_RETURN`:
  `pwsh`:
  ```pwsh
  "`e[20h"
  "test: line1`nline2`nline3"
  
  ```

**Reply by kelleyma49** [reply]
@o-sdn-o , your post was very useful! Running `./con_mode.ps1 0` did undo the issue in a terminal session. I am unblocked!
